test/decl/protocol/protocols_with_self_or_assoc_reqs.swift

```diff
@@ -0,0 +1,128 @@
+}
+
+let p1: P1 = S1()
+_ = p1.returnSelf() // ok
+_ = p1.returnAssoc() // expected-error {{member 'returnAssoc' cannot be used on value of protocol type 'P1'; use a generic constraint instead}}
+
+func takesP1(arg: P1) {
+  _ = arg.returnSelf() // ok
+  _ = p1.returnAssoc() // expected-error {{member 'returnAssoc' cannot be used on value of protocol type 'P1'; use a generic constraint instead}}
+}
+
+takesP1(arg: p1) // ok
+
+let p1Array: [P1] = [S1()] // ok
+p1Array.forEach { 
+  _ = $0.returnSelf() // ok
+  _ = $0.returnAssoc() // expected-error {{member 'returnAssoc' cannot be used on value of protocol type 'P1'; use a generic constraint instead}}
+}
+
+protocol P2 {
+  associatedtype Q
+  func takesSelf(_: Self)
+  func takesAssoc(_: Q)
+  func takesNestedSelf(closure: (Self) -> ())
+  func takesNestedAssoc(closure: (Q) -> ())
+}
+
+struct S2: P2 {
+  typealias Q = Int
+  func takesSelf(_: Self) {}
+  func takesAssoc(_: Q) {}
+  func takesNestedSelf(closure: (Self) -> ()) { print(closure(S2())) }
+  func takesNestedAssoc(closure: (Q) -> ()) { print(closure(0)) }
+}
+
+let p2: P2 = S2()
+p2.takesSelf(S2()) // expected-error {{member 'takesSelf' cannot be used on value of protocol type 'P2'; use a generic constraint instead}}
+p2.takesAssoc(0) 
+// expected-error@-1 {{member 'takesAssoc' cannot be used on value of protocol type 'P2'; use a generic constraint instead}} 
+// expected-error@-2 {{cannot convert value of type 'Int' to expected argument type 'P2.Q'}}
+p2.takesNestedSelf { _ in } // okay
+p2.takesNestedAssoc { _ in } 
+// expected-error@-1 {{member 'takesNestedAssoc' cannot be used on value of protocol type 'P2'; use a generic constraint instead}}
+// expected-error@-2 {{cannot convert value of type '(_) -> ()' to expected argument type '(P2.Q) -> ()'}}
+
+func takesP2(arg: P2) {
+  arg.takesSelf(S2()) // expected-error {{member 'takesSelf' cannot be used on value of protocol type 'P2'; use a generic constraint instead}}
+  arg.takesAssoc(0) 
+  // expected-error@-1 {{member 'takesAssoc' cannot be used on value of protocol type 'P2'; use a generic constraint instead}} 
+  // expected-error@-2 {{cannot convert value of type 'Int' to expected argument type 'P2.Q'}}
+  arg.takesNestedSelf { _ in } // okay
+  arg.takesNestedAssoc { _ in } 
+  // expected-error@-1 {{member 'takesNestedAssoc' cannot be used on value of protocol type 'P2'; use a generic constraint instead}}
+  // expected-error@-2 {{cannot convert value of type '(_) -> ()' to expected argument type '(P2.Q) -> ()'}}
+}
+
+takesP2(arg: p2) // okay
+
+protocol P3 {
+  associatedtype Q
+  var assocProp: Q { get }
+  subscript(q: Q) -> Q { get }
+  var selfProp: Self { get }
+}
+
+struct S3: P3 {
+  typealias Q = Int
+  var assocProp: Q { 0 }
+  subscript(q: Q) -> Q { 0 }
+  var selfProp: Self { self }
+}
+
+let p3: P3 = S3()
+_ = p3.assocProp // expected-error {{member 'assocProp' cannot be used on value of protocol type 'P3'; use a generic constraint instead}}
+_ = p3[0]
```

Comment:
WDYT about using function parameters or smth else to avoid defining conforming types and keep the tests simple? We only really need concrete conformances in the execution test, I think.

Reply:
Yeah, good point. I’ll do that.